{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e9d71acc-d2b1-4f5e-9806-edf14ae61bd0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(rtracklayer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "88facc24-65b4-45c7-be24-cb97cbedd437",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#FOR WHEN STARTING WITH SOME FILTERED SNPS FROM NOSE-JAW scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c107eb87-e1c4-4420-b5fa-1fb68fe4aaba",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "snp_df <- read.csv('white_SNPs_filtered_nose_jaw_scores.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8d30c337-1923-4870-8bee-c7eff0711074",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Suppose your SNP dataframe is called snp_df, and column CHR_POS_RS looks like \"1_1234567_rs1234\"\n",
    "snp_df$chr <- sub(\"^(\\\\d+)_.*\", \"\\\\1\", snp_df$CHR_POS_RS)\n",
    "snp_df$pos <- as.integer(sub(\"^\\\\d+_(\\\\d+)_.*\", \"\\\\1\", snp_df$CHR_POS_RS))\n",
    "snp_df$rsid <- sub(\"^\\\\d+_\\\\d+_(.*)$\", \"\\\\1\", snp_df$CHR_POS_RS)\n",
    "\n",
    "# Ensure chromosome names match UCSC format\n",
    "snp_df$chr <- paste0(\"chr\", snp_df$chr)\n",
    "\n",
    "# Make GRanges (SNPs are 1bp intervals)\n",
    "snp_gr <- GRanges(\n",
    "  seqnames = snp_df$chr,\n",
    "  ranges = IRanges(start = snp_df$pos, end = snp_df$pos),\n",
    "  rsid = snp_df$rsid\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d608855a-0b2a-46fa-bcd8-5c66d7c3ef78",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "setwd('/home/alek/FACE/DATA/GWAS/white_summary_stats')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9b4e5e65-4138-4fc9-a5d1-af1272c8eea4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load the chain file if needed\n",
    "#chain <- import.chain(\"hg19ToHg38.over.chain\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4620a5fa-52cf-45cb-8199-fa4748cd3fc8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Discarding unchained sequences: chr23\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Liftover\n",
    "snp_gr_hg38 <- liftOver(snp_gr, chain)\n",
    "\n",
    "# Keep only mappings with exactly 1 result\n",
    "snp_gr_hg38 <- unlist(snp_gr_hg38[lengths(snp_gr_hg38) == 1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cca6f177-1582-4263-a270-2550c3e56020",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "snp_hg38_df <- data.frame(\n",
    "  chr = as.character(seqnames(snp_gr_hg38)),\n",
    "  pos = start(snp_gr_hg38),\n",
    "  rsid = snp_gr_hg38$rsid\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "faa5b261-3b6b-43c4-a1cc-aeef3d68bb24",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Merge hg38 coordinates into your original snp_df by 'rsid'\n",
    "snp_df_hg38 <- merge(\n",
    "  snp_df,\n",
    "  snp_hg38_df,\n",
    "  by = \"rsid\",\n",
    "  all.x = TRUE,   # keep all SNPs from your original file\n",
    "  suffixes = c(\"_hg19\", \"_hg38\")\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "617adc60-7a15-4810-bd54-a08d3aa5442f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "snp_df_hg38$chr_hg38 <- sub(\"^chr\", \"\", snp_df_hg38$chr_hg38)\n",
    "\n",
    "# Optionally, create a new CHR_POS_RS column for hg38\n",
    "snp_df_hg38$CHR_POS_RS_hg38 <- paste(snp_df_hg38$chr_hg38, snp_df_hg38$pos_hg38, snp_df_hg38$rsid, sep = \"_\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "40bb987e-a6e1-49a1-ab6d-16b3d0bf2096",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "snp_df_hg38$CHR_POS_RS <- snp_df_hg38$CHR_POS_RS_hg38\n",
    "snp_df_hg38$CHR_POS_RS_hg38 <- NULL\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "77a7c423-c87d-46dc-80ef-7c6b066465ed",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "write.csv(snp_df_hg38, 'white_SNPs_filtered_nose_jaw_scores_hg38.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a72c1d03-1cea-488c-9152-0c1f13ffe720",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "883edb89-eace-4447-afd6-cb68a85ee2b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ANOTHER FILE NOW , FOR WHEN STARTING WITH RAW WHITE SNPs\n",
    "snp_df1 <- read.csv('pvalues_MinP_filtered.csv')\n",
    "snp_df2 <- read.csv('snp_info_filtered.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "c69ef5b3-8ebf-4e53-a1fa-38f3ea93e708",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "snp_df <- cbind(snp_df2, snp_df1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5b4d1f3a-884f-46f5-bc2f-34d7188f4369",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "# Ensure chromosome names match UCSC format\n",
    "#snp_df$CHR <- paste0(\"chr\", snp_df$CHR)\n",
    "\n",
    "#DEPENDS ON THE FORMAT OF THE INPUT\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "2ddd7052-9dc0-4cc3-846f-2a3c18c60bad",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Discarding unchained sequences: chr23\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Make GRanges (SNPs are 1bp intervals)\n",
    "snp_gr <- GRanges(\n",
    "  seqnames = snp_df$CHR,\n",
    "  ranges = IRanges(start = snp_df$POS, end = snp_df$POS),\n",
    "  rsid = snp_df$RS\n",
    ")\n",
    "\n",
    "setwd('/home/alek/FACE/DATA/GWAS/white_summary_stats')\n",
    "\n",
    "chain <- import.chain(\"hg19ToHg38.over.chain\")\n",
    "\n",
    "snp_gr_hg38 <- liftOver(snp_gr, chain)\n",
    "snp_gr_hg38 <- unlist(snp_gr_hg38[lengths(snp_gr_hg38) == 1])\n",
    "\n",
    "\n",
    "snp_hg38_df <- data.frame(\n",
    "  chr = as.character(seqnames(snp_gr_hg38)),\n",
    "  pos = start(snp_gr_hg38),\n",
    "  rsid = snp_gr_hg38$rsid\n",
    ")\n",
    "\n",
    "if(!\"rsid\" %in% colnames(snp_df)) colnames(snp_df)[colnames(snp_df) == \"RS\"] <- \"rsid\"\n",
    "\n",
    "snp_df_hg38 <- merge(\n",
    "  snp_df,\n",
    "  snp_hg38_df,\n",
    "  by = \"rsid\",\n",
    "  all.x = TRUE,   # keep all SNPs from your original file\n",
    "  suffixes = c(\"_hg19\", \"_hg38\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "01bee351-9ef1-4c9e-b7ad-c22c7b8d4483",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "colnames(snp_df_hg38)[colnames(snp_df_hg38) == \"POS\"] <- \"POS_hg19\"\n",
    "colnames(snp_df_hg38)[colnames(snp_df_hg38) == \"pos\"] <- \"pos_hg38\"\n",
    "colnames(snp_df_hg38)[colnames(snp_df_hg38) == \"chr\"] <- \"chr_hg38\"\n",
    "colnames(snp_df_hg38)[colnames(snp_df_hg38) == \"CHR\"] <- \"CHR_hg19\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "3d10d27d-b0ab-4400-90f7-a5e846469f99",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "write.csv(snp_df_hg38, 'WHITE_hg38.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03ae2c5e-29a3-4299-a622-640cf32ba2f8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "AlekR",
   "language": "R",
   "name": "alekr"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
